index.go

```diff
@@ -585,3 +611,113 @@ type importData struct {
 	RowIDs    []uint64
 	ColumnIDs []uint64
 }
+
+// CreateInputDefinition creates a new input definition.
+func (i *Index) CreateInputDefinition(pb *internal.InputDefinition) (*InputDefinition, error) {
+	// Ensure input definition doesn't already exist.
+	if i.inputDefinitions[pb.Name] != nil {
+		return nil, ErrInputDefinitionExists
+	}
+	return i.createInputDefinition(pb)
+}
+
+func (i *Index) createInputDefinition(pb *internal.InputDefinition) (*InputDefinition, error) {
+	if pb.Name == "" {
+		return nil, errors.New("input-definition name required")
+	} else if len(pb.Frames) == 0 || len(pb.Fields) == 0 {
+		return nil, errors.New("frames and fields are required")
+	}
+
+	for _, fr := range pb.Frames {
+		opt := FrameOptions{
+			RowLabel:       fr.Meta.RowLabel,
+			InverseEnabled: fr.Meta.InverseEnabled,
+			CacheType:      fr.Meta.CacheType,
+			CacheSize:      fr.Meta.CacheSize,
+			TimeQuantum:    TimeQuantum(fr.Meta.TimeQuantum),
+		}
+		_, err := i.CreateFrame(fr.Name, opt)
+		if err == ErrFrameExists {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+	}
+
+	// Initialize input definition.
+	inputDef, err := i.newInputDefinition(i.InputDefPath(), pb.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	inputDef.LoadDefinition(pb)
+	if err = inputDef.saveMeta(); err != nil {
+		return nil, err
+	}
+	i.inputDefinitions[pb.Name] = inputDef
+	return inputDef, nil
+}
+
+func (i *Index) newInputDefinition(path, name string) (*InputDefinition, error) {
```

Comment:
will there ever be a case where `path` is not `i.InputDefPath()`? if not, then we can remove the `path` argument from this function signature and just replace `i.name` with `i.InputDefPath()` inside the function.